src/isEmpty.ts

```diff
@@ -0,0 +1,39 @@
+import { isArray } from './isArray';
+import { isObject } from './isObject';
+import { isString } from './isString';
+
+type Data =
+  | ReadonlySet<unknown>
+  | ReadonlyMap<unknown, unknown>
+  | ArrayLike<unknown>
+  | string
+  | Record<string, unknown>;
+
+/**
+ * A function that checks if the passed parameter is empty
+ * @param data the variable to check
+ * @signature
+ *    R.isEmpty(data)
+ * @returns true if the passed input is empty, false otherwise
+ * @example
+ *    R.isEmpty([]) //=> true
+ *    R.isEmpty({}) //=> true
+ *    R.isEmpty('') //=> true
+ *    R.isEmpty(new Set()) //=> true
+ *    R.isEmpty(new Map()) //=> true
+ *    R.isEmpty([1, 2, 3]) //=> false
+ *    R.isEmpty({ length: 0 }) //=> false
+ * @category Function
+ */
+export function isEmpty(data: Data): boolean {
```

Comment:
Thanks for handling the typing, I still think your original approach to narrow the type too is still valid here, as any data refinement is better than none. You can type the return value based on T so that you can "link" the input and the output.

With this kind of typing you don't need to type the input itself, because any other value (like number) would fall to the `never` final type, and the code wouldn't compile. (you can test this with @ts-expect-error above calls to `isEmpty(3)`)

```suggestion
type Empty<T> = T extends ReadonlySet<unknown>
  ? ReadonlySet<never>
  : T extends ReadonlyMap<infer Key, unknown>
  ? ReadonlyMap<Key, never>
  : T extends ArrayLike<unknown>
  ? ArrayLike<never>
  : T extends string
  ? ""
  : T extends Record<PropertyKey, unknown>
  ? Record<keyof T, never>
  : never; 

export function isEmpty<T>(data: T): data is Empty<T> {
```

Reply:
Hmm, I like the idea, but I'm not exactly sure how is it possible to link the input and the output, since, for `data is Empty<T>`  type predicate, Typescript complains that: 
```ts
A type predicate's type must be assignable to its parameter's type.
  Type 'Empty<T>' is not assignable to type 'T'.
    'T' could be instantiated with an arbitrary type which could be unrelated to 'Empty<T>'.
```

Reply:
I believe the below code works as intended, according to what you have described. 

Unfortunately, I'm not using `readonly` types here – I still need to work on them a little bit to avoid them being narrowed down to `never` when the `data` argument is empty, instead of, for example, `ReadonlySet<never>`.

```ts
type Empty<T> = T extends Map<infer Key, unknown>
  ? Map<Key, never>
  : T extends Set<unknown>
  ? Set<never>
  : T extends string
  ? ''
  : T extends Array<unknown>
  ? Array<never>
  : T extends Record<PropertyKey, unknown>
  ? Record<keyof T, never>
  : never;

export function isEmpty<T>(
  data: Empty<T> extends never ? never : T
): data is Empty<T> extends never
  ? never
  : Empty<T> extends T
  ? Empty<T>
  : never {
  if (data instanceof Set || data instanceof Map) {
    return data.size === 0;
  }
  if (isArray(data) || isString(data)) {
    return data.length === 0;
  }
  if (isObject(data)) {
    return Object.keys(data).length === 0;
  }
  return false;
}

```

Tests of narrowing:
```ts
// ✅
let input = 'test';
if (isEmpty(input)) {
  input; //=> ""
} else {
  input; //=> string
}

// ✅
const input1: Array<number> = [];
if (isEmpty(input1)) {
  input1; //=> never[]
} else {
  input1; //=> number[]
}

// ✅
const input2: Record<string, unknown> = {};
if (isEmpty(input2)) {
  input2; //=> Record<string, never>
} else {
  input2; //=> Record<string, unknown>
}

// ✅
const input3 = { a: 1, b: 2 };
if (isEmpty(input3)) {
  input3; //=> Record<"a" | "b", never>
} else {
  input3; // => {
  //      a: number;
  //      b: number;
  //   }
}

// ✅
const input4 = new Set(['a', 2]);
if (isEmpty(input4)) {
  input4; // => Set<never>
} else {
  input4; // => Set<string | number>
}

// ✅
const input5 = 23;
isEmpty(input5); // => Error: Argument of type number is not assignable to parameter of type never.
```

Reply:
Sorry for driving you crazy with this, working on Remeda is always a good way to practice typescript :)

I'm fine with dropping the readonly parts, maybe except the readonly Array which I know is very predictable.

Another approach is to simply use function overloading here, it might just be easier to read and maybe work better. Typescript would do the complex union typing for us via the overload.

Lastly, let's make arrays return `[]` and not `Array<never>`, i'm concerned that `Array<never>` can still be a sparse array with length !== 0, `[]` is fully unambiguous.

PS, i'm not sure that type-narrowing Sets and Maps is useful anyway, i'm not sure they have a good semantic of an empty collection as they would suffer from the same problem I mentioned with the array. I'm also not sure they are used frequently enough that we should care right now. We can always fix bugs that get reported later on.

```ts
export function isEmpty(data: string): data is "";
export function isEmpty(data: ReadonlyArray<unknown> | []): data is [];
export function isEmpty<Key extends PropertyKey>(data: Readonly<Record<Key, unknown>>): data is Record<Key, never>;
export function isEmpty(data: ReadonlySet<unknown>): data is Set<never>;
export function isEmpty<Key>)(data: ReadonlyMap<Key, unknown>): data is Map<Key, never>;
export function isEmpty(data: unknown): boolean {
  // IMPL
}
```

Reply:
I've already learned a ton, so I really appreciate it! :) 

I pushed the changes that include function overloading, just as you've suggested and it make the code much more readable, that's for sure. I have only slightly changed a signature of `export function isEmpty<Key extends PropertyKey>(data: Readonly<Record<Key, unknown>>): data is Record<Key, never>`, because it made the function accept invalid input without an error. 
I have also removed type-narrowing for Sets and Maps – I agree with you that they might not be that relevant but if they turn out to be so, then we can always fix that later.